Add RFC for improving client error ergonomics (#2075)
smithy-lang/smithy-rs: design/src/rfcs/rfc0026_improve_client_error_ergonomics.md

With the SDK and code generated clients, customers have two primary use-cases that should be made
easy without compromising the compatibility rules established in [RFC-0022]:

1. Checking the error type
2. Retrieving information specific to that error type

Case Study: Handling an error in S3
-----------------------------------

The following is an example of handling errors with S3 with the latest generated (and unreleased)
SDK as of 2022-12-07:

```rust
let result = client
    .get_object()
    .bucket(BUCKET_NAME)
    .key("some-key")
    .send()
    .await;
match result {
    Ok(_output) => { /* Do something with the output */ }
    Err(err) => match err.into_service_error() {
        GetObjectError {
            kind: GetObjectErrorKind::InvalidObjectState(value),
            ..
        } => {
            println!("invalid object state: {:?}", value);
        }
        GetObjectError {
            kind: GetObjectErrorKind::NoSuchKey(_),
            ..
        } => {
            println!("object didn't exist");
        }
        err @ GetObjectError { .. } if err.code() == Some("SomeUnmodeledError") => {}
        err @ _ => return Err(err.into()),
    },

> }

Author: Nice!

```

The refactor that implemented [RFC-0022] added the `into_service_error()` method on `SdkError` that

> infallibly converts the `SdkError` into the generic error type for the `SdkError::ServiceError` variant.

Reviewer: ```suggestion
infallibly converts the `SdkError` into the concrete error type held by the `SdkError::ServiceError` variant.
```

> This improvement let customers discard the transient failures to immediately handle modeled errors
> returned by the service.

Reviewer: Nit. I prefer present tense for some reason.
```suggestion
This improvement lets customers discard transient failures and immediately handle modeled errors
returned by the service.
```

Proposal: Combine `Error` and `ErrorKind`
-----------------------------------------

At time of writing, each operation has both an `Error` and `ErrorKind` type generated.
The `Error` type holds information that is common across all operation errors: message,
error code, "extra" key/value pairs, and the request ID.

The `ErrorKind` is always nested inside the `Error`, which results in the verbose
nested matching shown in the case study above.

To make error handling more ergonomic, the code generated `Error` and `ErrorKind` types
should be combined. Hypothetically, this would allow for the case study above to look as follows:

```rust
let result = client
    .get_object()
    .bucket(BUCKET_NAME)
    .key("some-key")
    .send()
    .await;
match result {
    Ok(_output) => { /* Do something with the output */ }
    Err(err) => match err.into_service_error() {
        GetObjectError::InvalidObjectState(value) => {
            println!("invalid object state: {:?}", value);
        }
        err if err.is_no_such_key() => {
            println!("object didn't exist");
        }
        err if err.code() == Some("SomeUnmodeledError") => {}
        err @ _ => return Err(err.into()),

>     },

Reviewer: This looks like a great UX! I'd enjoy interacting with this.

}
```

If a customer only cares about checking one specific error type, they can also do:

```rust
match result {
    Ok(_output) => { /* Do something with the output */ }
    Err(err) => {
        let err = err.into_service_error();
        if err.is_no_such_key() {
            println!("object didn't exist");
        } else {
            return Err(err);
        }
    }
}
```

The downside of this is that combining the error types requires adding the general error
metadata to each generated error struct so that it's accessible by the enum error type.
However, this aligns with our tenet of making things easier for customers even if it

> makes it harder for ourselves.

Reviewer: This makes me really wish that Rust had delegation. Then we'd only have to write the common part once.
